### PR TITLE
WIP: Adds WDL support and other fixes

### PR DESCRIPTION
**Pull request by scanon**

### CODE REVIEW

File: JobRunner/SpecialRunner.py [hunk 1]
```diff
@@ -58,23 +62,27 @@ def _watch_batch(self, stype, job_id, slurm_jobid,
         self.logger.log("Watching Slurm Job ID %s" % (slurm_jobid))
         check = '%s_checkjob' % (stype)
         cont = True
-        started = False
+        retry = 0
```
Comment (bio-boris) [inline JobRunner/SpecialRunner.py]: 10 seconds of waiting with 5 retries? This looks like a very short time to let a job run. Is this intended?

Comment (bio-boris) [reply JobRunner/SpecialRunner.py]: Oh wait, if its running, the retries are not incremented, never mind. Its 50 seconds of lookup failures?

File: test/test_jobrunner.py [hunk 2]
```diff
@@ -357,3 +357,30 @@ def test_run_slurm(self, mock_njs, mock_auth):
...
+                       self.admin_token)
+        mlog = MockLogger()
+        jr.logger = mlog
+        jr.sr.logger = mlog
+        rv = deepcopy(CATALOG_GET_MODULE_VERSION)
+        rv['docker_img_name'] = 'test/runtester:latest'
+        jr.cc.catalog.get_module_version = MagicMock(return_value=rv)
+        jr.cc.catalog.list_volume_mounts = MagicMock(return_value=[])
+        jr.cc.catalog.get_secure_config_params = MagicMock(return_value=None)
+        jr.njs.get_job_params.return_value = params
```
Comment (bio-boris) [inline test/test_jobrunner.py]: need to set    jr.njs.get_job_params.return_value = params[0]
need to set list_config() = params[1]

File: JobRunner/SpecialRunner.py [hunk 3]
```diff
@@ -141,3 +158,60 @@ def _batch_submit(self, stype, config, data, job_id, fin_q):
         out.start()
         self.containers.append(proc)
         return proc
+
+    def _readio(self, p, job_id, queues):
```
Comment (bio-boris) [inline JobRunner/SpecialRunner.py]: This function is a bit confusing. Some clearer variable names would be helpful

File: JobRunner/MethodRunner.py [hunk 4]
```diff
@@ -136,7 +136,8 @@ def run(self, config, module_info, params, job_id, fin_q=None,
             job_dir: {'bind': '/kb/module/work', 'mode': 'rw'}
         }
         if subjob:
-            wdt = os.path.join(job_dir, "/tmp")
+            parent_workdir = self._get_job_dir(job_id, subjob=False)
+            wdt = os.path.join(parent_workdir, "tmp")
```
Comment (bio-boris) [inline JobRunner/MethodRunner.py]: wdt (workdirtemp?) is a cryptic variable name

File: JobRunner/DockerRunner.py [hunk 5]
```diff
@@ -71,20 +71,21 @@ def _shepherd(self, c, job_id, queues):
...
                     dolast = True
+        except Exception as e:
+            if self.logger is not None:
+                self.logger.error("Unexpected failure")
+            else:
+                print("Exception in docker logging for %s" % (c.id))
+                raise(e)
+        finally:
             try:
                 c.remove()
```
Comment (bio-boris) [inline JobRunner/DockerRunner.py]: Maybe the function dockerruner.remove() should be used?

**Comment by scanon** [comment]
I think this is deprecated

**Comment by bio-boris** [comment]
Fine by me to close it

**Comment by scanon** [comment]
Merged by other branch

**Pull request closed without merging**